libc/string_mem.c

```diff
@@ -0,0 +1,396 @@
+/*
+ * Phoenix-RTOS
+ *
+ *    libc-tests
+ *    HEADER:
+ *    - string.h
+ *    TESTED:
+ *    - memset(),
+ *    - memmove()
+ *
+ * Copyright 2023 Phoenix Systems
+ * Author: Adam Debek
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <string.h>
+#include <unistd.h>
+#include <stdlib.h>
+#include <errno.h>
+#include <ulimit.h>
+#include <stdint.h>
+#include <unity_fixture.h>
+
+#define BUF_LEN         16
+#define VALUES_NUM      sizeof(values) / sizeof(values[0])
+#define BIG_SIZE        1024
+#define TEST_STR        "THIS IS TEST MEM123"
+#define TEST_STR_SIZE   sizeof(TEST_STR)
+#define OFFSET          40
+#define OFFSET_BIG      1500
+#define MEM_SIZE        100
+#define MEM_SIZE_BIG    4000
+#define OVERLAP_LEN     (TEST_STR_SIZE / 2)
+#define OVERLAP_LEN_BIG (BIG_SIZE / 2)
+
+static uint8_t buf[BUF_LEN];
+static uint8_t buf_byte[256];
+static uint8_t buf_byte_exp[256];
+static uint8_t *ret;
+static uint8_t *mem;
+static uint8_t *mem_big;
+static uint8_t big_buf[1024];
+static uint8_t expect_mem_big[BIG_SIZE];
+static int values[] = { 0, 1, 16, 50, 127, 128, 200, 255,
+	768, 257, 272, 306, 383, 384, 456, 511 };
+/* next values exceed 1 byte, (uint8_t)385 = -127, (uint8_t)511 = -1 and so on */
+
+
+TEST_GROUP(string_memset);
+TEST_GROUP(string_memmove);
+TEST_GROUP(string_memmove_big);
+
+
+TEST_SETUP(string_memset)
+{
+}
+
+
+TEST_TEAR_DOWN(string_memset)
+{
+}
+
+
+TEST(string_memset, basic)
+{
+	int i;
+	for (i = 0; i < VALUES_NUM; i++) {
+		ret = memset(buf, values[i], BUF_LEN);
+		TEST_ASSERT_EQUAL_PTR(buf, ret);
+		TEST_ASSERT_EACH_EQUAL_INT8((uint8_t)values[i], buf, BUF_LEN);
+	}
+}
+
+
+TEST(string_memset, zero)
+{
+	uint8_t val = 5;
+	ret = memset(&val, 0, 0);
+	TEST_ASSERT_EQUAL_PTR(&val, ret);
+	TEST_ASSERT_EQUAL_INT8(5, val);
+}
+
+
+TEST(string_memset, big)
+{
+	ret = memset(big_buf, 10, BIG_SIZE);
+	TEST_ASSERT_EQUAL_PTR(big_buf, ret);
+	TEST_ASSERT_EQUAL_MEMORY(expect_mem_big, big_buf, BIG_SIZE);
```

Comment:
use `EACH_EQUAL_UINT8` and remove `expect_mem_big`

**Important:**
the `memset` and `memcpy` tests should test lengths and starting pointers not dividable by 2, 4, 8  (so with `len % 8` and `ptr % 8` with all combinations) - common optimization is to use full words for memset / memcpy if possible

Reply:
Do you mean to add test case with odd ptr and len?